src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1547,3 +2145,348 @@ MultiRouterPlannableQuery(Query *query, MultiExecutorType taskExecutorType,
+										PVC_RECURSE_PLACEHOLDERS);
+		targetVarCount = list_length(targetVarList);
+
+		/* a single INSERT target entry cannot have more than one Var */
+		Assert(targetVarCount <= 1);
+
+		if (targetVarCount == 1)
+		{
+			Var *oldInsertVar = (Var *) linitial(targetVarList);
+			TargetEntry *oldSubqueryTle = list_nth(subquery->targetList,
+												   oldInsertVar->varattno - 1);
+
+			newSubqueryTargetEntry = copyObject(oldSubqueryTle);
+
+			newSubqueryTargetEntry->resno = resno;
+			newSubqueryTargetlist = lappend(newSubqueryTargetlist,
+											newSubqueryTargetEntry);
+
+			updatedSubqueryEntryCount++;
+		}
+		else
+		{
+			newSubqueryTargetEntry = makeTargetEntry(oldInsertTargetEntry->expr,
+													 resno,
+													 oldInsertTargetEntry->resname,
+													 oldInsertTargetEntry->resjunk);
+			newSubqueryTargetlist = lappend(newSubqueryTargetlist,
+											newSubqueryTargetEntry);
+		}
+
+		newInsertVar = makeVar(insertTableId, originalAttrNo,
+							   exprType((Node *) newSubqueryTargetEntry->expr),
+							   exprTypmod((Node *) newSubqueryTargetEntry->expr),
+							   exprCollation((Node *) newSubqueryTargetEntry->expr),
+							   0);
+		newInsertTargetEntry = makeTargetEntry((Expr *) newInsertVar, originalAttrNo,
+											   oldInsertTargetEntry->resname,
+											   oldInsertTargetEntry->resjunk);
+
+		newInsertTargetlist = lappend(newInsertTargetlist, newInsertTargetEntry);
+		resno++;
+	}
+
+	/*
+	 * if there are any remaining target list entries (i.e., GROUP BY column not on the
+	 * target list of subquery), update the remaining resnos.
+	 */
+	subqueryTargetLength = list_length(subquery->targetList);
+	if (subqueryTargetLength != updatedSubqueryEntryCount)
+	{
+		int targetEntryIndex = updatedSubqueryEntryCount;
+
+		for (; targetEntryIndex < subqueryTargetLength; ++targetEntryIndex)
+		{
+			TargetEntry *oldSubqueryTle = list_nth(subquery->targetList,
+												   targetEntryIndex);
+			TargetEntry *newSubqueryTargetEntry = copyObject(oldSubqueryTle);
+
+			Assert(newSubqueryTargetEntry->resjunk == true);
+
+			newSubqueryTargetEntry->resno = resno;
+			newSubqueryTargetlist = lappend(newSubqueryTargetlist,
+											newSubqueryTargetEntry);
+
+			resno++;
+		}
+	}
+
+	originalQuery->targetList = newInsertTargetlist;
+	subquery->targetList = newSubqueryTargetlist;
+
+	return NULL;
+}
+
+
+/*
+ * InsertSelectQuery returns true when the input query
+ * is INSERT INTO ... SELECT kind of query.
+ */
+bool
+InsertSelectQuery(Query *query)
+{
+	CmdType commandType = query->commandType;
+	List *rangeTableList = query->rtable;
+	RangeTblEntry *subqueryRte = NULL;
+	Query *subquery = NULL;
+
+	if (commandType != CMD_INSERT)
+	{
+		return false;
+	}
+
+	rangeTableList = query->rtable;
+	if (list_length(rangeTableList) < 2)
+	{
+		return false;
+	}
+
+	subqueryRte = lsecond(query->rtable);
+	subquery = subqueryRte->subquery;
+	if (subquery == NULL)
+	{
+		return false;
+	}
```

Comment:
Using the rtable count isn't a good idea. There's a number of cases where stuff can end up the rtable without having any relevance on the query.  Looking at ->jointree seems a lot more reasonable.